patches/minecraft/net/minecraft/world/server/ServerWorld.java.patch

```diff
@@ -306,7 +306,45 @@
        if (p_217401_2_ != null) {
           explosion.func_199592_a(p_217401_2_);
        }
-@@ -1377,4 +1421,15 @@
+@@ -1109,22 +1153,25 @@
+    }
+ 
+    public <T extends IParticleData> int func_195598_a(T p_195598_1_, double p_195598_2_, double p_195598_4_, double p_195598_6_, int p_195598_8_, double p_195598_9_, double p_195598_11_, double p_195598_13_, double p_195598_15_) {
+-      SSpawnParticlePacket sspawnparticlepacket = new SSpawnParticlePacket(p_195598_1_, false, p_195598_2_, p_195598_4_, p_195598_6_, (float)p_195598_9_, (float)p_195598_11_, (float)p_195598_13_, (float)p_195598_15_, p_195598_8_);
+-      int i = 0;
+-
+-      for(int j = 0; j < this.field_217491_A.size(); ++j) {
+-         ServerPlayerEntity serverplayerentity = this.field_217491_A.get(j);
+-         if (this.func_195601_a(serverplayerentity, false, p_195598_2_, p_195598_4_, p_195598_6_, sspawnparticlepacket)) {
+-            ++i;
++      java.util.concurrent.atomic.AtomicInteger i = new java.util.concurrent.atomic.AtomicInteger();
++      net.minecraftforge.event.ForgeEventFactory.onWorldParticleSpawn(p_195598_1_, false, p_195598_2_, p_195598_4_, p_195598_6_, p_195598_8_, p_195598_9_, p_195598_11_, p_195598_13_, p_195598_15_).ifPresent(event -> {
++         SSpawnParticlePacket sspawnparticlepacket = new SSpawnParticlePacket(event.getParticle(), event.isLongDistance(), (float)event.getXCoord(), (float)event.getYCoord(), (float)event.getZCoord(), (float)event.getXOffset(), (float)event.getYOffset(), (float)event.getZOffset(), (float)event.getParticleSpeed(), event.getParticleCount());
++         for (ServerPlayerEntity serverplayerentity : this.field_217491_A) {
++            if (this.func_195601_a(serverplayerentity, event.isLongDistance(), event.getXCoord(), event.getYCoord(), event.getZCoord(), sspawnparticlepacket)) {
++               i.incrementAndGet();
++            }
+          }
+-      }
+-
+-      return i;
++      });
++      return i.get();
```

Comment:
You can make this patch _way_ smaller by not using `Optional`. Make `onParticleSpawn` return the event, or null. Then just have one if statement that returns immediately if the event is cancelled (two lines so far). After that just re-assign the parameter variables using the event. Then your patch becomes only additions, which is much easier to maintain.

Reply:
@ChristianANoeddeboe sorry to indeed have send you on a goose chase, i misread the PR and the patch. Using an Optional here is indeed not usefull. But you will need to have the method in ForgeEventFactory @diesieben07 is completely right here.

The best approach would probably be to combine what you had in the first place with the factory method either returning null to cancel the behaviour, or returning the processed event. This would indeed make it a two, possibly three, line patch.

Sorry for the inconvenience.